### PR TITLE
Carousel Improvements

### PR DESCRIPTION
**Pull request by tobi-or-not-tobi**
Draft for all the changes of #9481

### CODE REVIEW

File: projects/storefrontlib/src/shared/components/carousel/carousel.component.ts [hunk 1]
```diff
@@ -62,32 +82,204 @@ export class CarouselComponent implements OnInit {
...
+  }
+
+  /**
+   * Set's the freeze class on the host element. This class is used in CSS to block
+   * interaction (`pointer-event: none`) with the panel, as otherwise a mouse hover
+   * would destroy the scroll animation.
+   */
+  protected freeze() {
+    this._freeze = true;
+    window.setTimeout(() => {
```
Comment (LTiger14) [inline projects/storefrontlib/src/shared/components/carousel/carousel.component.ts]: Please use windowRef

File: projects/storefrontlib/src/shared/components/carousel/carousel.component.ts [hunk 2]
```diff
@@ -38,11 +44,25 @@ export class CarouselComponent implements OnInit {
...
+   * provides a configuration to add accessibility control to the carousel.
+   * The default configuration adds the following:
+   * - the first item is auto-focused (if it is focusable) is selected.
+   *
+   * Additionally, a named focus _group_ could be added to refocus the last selected element
+   * if the carousel is focussed again.
+   */
+  @Input() focusConfig: FocusConfig = {
+    autofocus: true,
+    // focusOnScroll: true,
```
Comment (LTiger14) [inline projects/storefrontlib/src/shared/components/carousel/carousel.component.ts]: Cleanup

File: projects/storefrontlib/src/layout/loading/intersect.directive.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    )
+  );
+
+  constructor(
+    protected intersectionService: IntersectionService,
+    protected elementRef: ElementRef
+  ) {}
+
+  protected get config(): IntersectionOptions {
+    // TODO: read the default config from the intersectionService->LayoutConfig
```
Comment (LTiger14) [inline projects/storefrontlib/src/layout/loading/intersect.directive.ts:46]: You could introduce optional `threshold` to the config is you want to keep backwards compatibility.

File: projects/storefrontlib/src/layout/a11y/keyboard-focus/lock/lock-focus.directive.ts [hunk 4]
```diff
@@ -158,7 +158,7 @@ export class LockFocusDirective
 
   handleEscape(event: KeyboardEvent): void {
     if (this.shouldLock) {
-      this.service.clear(this.config.group);
+      // this.service.clear(this.config.group);
```
Comment (LTiger14) [inline projects/storefrontlib/src/layout/a11y/keyboard-focus/lock/lock-focus.directive.ts:161]: Why did you change this?

File: projects/storefrontlib/src/cms-components/product/carousel/product-carousel/product-carousel.component.html [hunk 5]
```diff
@@ -1,19 +1,40 @@
+<!-- *ngIf="(title$ | async) === 'Our Bestselling Products'" -->
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/product/carousel/product-carousel/product-carousel.component.html]: Cleanup

File: projects/storefrontlib/src/shared/components/carousel/carousel.model.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { ICON_TYPE } from '../../../cms-components/misc/icon/icon.model';
+
+export interface CarouselButton {
+  // visible: boolean;
```
Comment (LTiger14) [inline projects/storefrontlib/src/shared/components/carousel/carousel.model.ts:4]: I don't think we need the visible property.

File: projects/storefrontlib/src/shared/components/carousel/carousel.component.html [hunk 7]
```diff
@@ -1,61 +1,87 @@
...
+    (click)="handlePrevious()"
+    [disabled]="previousButton.disabled"
+    aria-label="Press to see previous slides"
+  >
+    <cx-icon [type]="previousButton.icon || previousIcon"></cx-icon>
+  </button>
+  <div
+    #carousel
+    class="carousel-panel"
+    [ngClass]="'size-' + size"
```
Comment (LTiger14) [inline projects/storefrontlib/src/shared/components/carousel/carousel.component.html]: Isn't the `size-...` class logic commented out of the css?

File: projects/storefrontlib/src/layout/a11y/keyboard-focus/tab/tab-focus.service.ts [hunk 8]
```diff
@@ -13,18 +13,19 @@ export class TabFocusService extends AutoFocusService {
...
-    event: KeyboardEvent
+    event?: KeyboardEvent
   ): void {
     if (config?.tab) {
       const next =
         config.tab === 'scroll'
           ? this.findNextScrollable(host, config, increment)
           : this.findNext(host, config, increment);
 
+      console.log('next', next);
```
Comment (LTiger14) [inline projects/storefrontlib/src/layout/a11y/keyboard-focus/tab/tab-focus.service.ts:24]: Cleanup

File: projects/storefrontlib/src/layout/a11y/keyboard-focus/autofocus/auto-focus.directive.ts [hunk 9]
```diff
@@ -50,6 +54,22 @@ export class AutoFocusDirective
     if (!this.shouldAutofocus || this.hasPersistedFocus) {
       super.ngAfterViewInit();
     }
+
+    if (this.config.focusOnScroll) {
+      // TODO: cleanup
```
Comment (LTiger14) [inline projects/storefrontlib/src/layout/a11y/keyboard-focus/autofocus/auto-focus.directive.ts]: Cleanup

File: projects/storefrontlib/src/shared/components/carousel/carousel.component.html [hunk 10]
```diff
@@ -1,61 +1,87 @@
...
-          <ng-container
-            *ngFor="let item of items | slice: i:i + size; let j = index"
+          <!-- 
+            We render the carousel item, so that we have an element to work with. This allows to do skeleton, prefetching and accessibility.
+          -->
+          <div
+            *ngFor="let data of items | slice: i:i + size; let itemNum = index"
+            class="item"
+            [class.active]="i === activeSlide"
+            (mouseenter)="prefetch()"
```
Comment (LTiger14) [inline projects/storefrontlib/src/shared/components/carousel/carousel.component.html]: Is this mouseenter required? I find it executes quite a bit of logic every time you hover an item. Obviously it won't send new network call when the data is already fetched but moving your mouse back and forth leads to a lot of function calls.

File: projects/storefrontstyles/scss/components/product/carousel/_carousel.scss [hunk 11]
```diff
@@ -1,118 +1,288 @@
 %cx-carousel {
```
Comment (LTiger14) [inline projects/storefrontstyles/scss/components/product/carousel/_carousel.scss:1]: CSS cleanup general comment 😉

**Review by LTiger14 — CHANGES_REQUESTED** [review]
In my opinion, the direction the carousel is heading into is very good. That said, there are a few areas I would like to improve.

1. The styling in Firefox and Safari needs a bit of tweaking. I found that despite the odd looks, the behaviour was still correct on both browsers. Additionally, the carousel didn't really like my Iphone 11 Pro simulator.

Firefox 82:
Screen Shot 2020-10-23 at 1 42 44 PM

Safari:
Screen Shot 2020-10-23 at 1 42 34 PM
Screen Shot 2020-10-23 at 1 42 38 PM

IOS safari, Iphone 11 PRO:
Screen Shot 2020-10-26 at 12 01 28 PM

2. I did some logging and I noticed that the `getProduct(code: string, prefetch: boolean)` in `product-carousel.component` got called a TON of times. I understand that there is some prefetching going on, I added a comment regarding the `mouseover` on carousel items that added a few calls.

3. The carousel has always been some pretty complicated code to understand but I feel like now it's gotten a step more difficult. I'm sorry I don't have a good suggestion to simplify it at the moment; perhaps if we get together we could find something.

**For after**: I added some cleanup comments so we don't forget.

File: projects/storefrontlib/src/shared/components/carousel/carousel.component.html (hunk 10 above)
Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/shared/components/carousel/carousel.component.html]: It is, but the current separate data we pass into the carousel item template is causing loads of change detection cycles. I believe we must change this into a single object with observables for the item, prefetch and tabindex. I need to test it out though.

**Comment by saad-mo** [comment]
Hello - @tobi-or-not-tobi @LTiger14, This PR contains a11y fixes for keyboard focus which would be helpful to have in place while we address Screen Reader issues. What is the likely hood of this getting done anytime soon?

File: projects/storefrontlib/src/shared/components/carousel/carousel.component.html (hunk 7 above)
Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/shared/components/carousel/carousel.component.html]: yes, it's gone now.

**Pull request closed without merging**